agentarchives/atom/client.py

```diff
@@ -0,0 +1,648 @@
+            # within that record with this one.
+            if not 'notes' in record or record['notes'] == []:
+                record['notes'] = [new_note]
+            else:
+                record['notes'][0] = new_note
+
+            fields_updated = True
+        else:
+            # Remove existing notes if the record didn't have a valid note;
+            # a note with an empty string as content should be counted as
+            # a request to delete the note.
+            record['notes'] = []
+
+        # Update date
+        updated_date = {}
+
+        # Only single dates are currently supported
+        if 'dates' in new_record and type(new_record['dates']) is list:
+            new_record['dates'] = new_record['dates'][0]
+
+        # Map agentarchives date specification to AtoM specification
+        date_mapping = {
+            'start_date': 'start_date',
+            #'begin': 'start_date',
+            'end_date': 'end_date',
+            #'end': 'end_date',
+            'date_expression': 'date'
+        }
+
+        for date_field in date_mapping:
+            if date_field in new_record:
+                updated_date[date_mapping[date_field]] = new_record[date_field]
+
+        # Add updated date specification to record update
+        if updated_date != {}:
+            record['dates'] = [updated_date]
+            fields_updated = True
+
+        if not fields_updated:
+            raise ValueError('No fields to update specified!')
+
+        self._put(self.api_base_url + '/informationobjects/' + record_id, data=json.dumps(record))
+
+    def get_levels_of_description(self):
+        """
+        Returns an array of all levels of description defined in this AtoM instance.
+        """
+        if not hasattr(self, 'levels_of_description'):
+            self.levels_of_description = [item['name'] for item in self._get(self.api_base_url + '/taxonomies/34').json()]
+
+        return self.levels_of_description
+
+    def collection_list(self, resource_id, resource_type='collection'):
+        """
+        Fetches a list of slug representing descriptions within the specified parent description.
+
+        :param resource_id str: The slug of the description to fetch children from.
+        :param resource_type str: no-op; not required or used in this implementation.
+
+        :return: A list of strings representing the slugs for all children of the requested description.
+        :rtype list:
+        """
+        def fetch_children(children):
+            results = []
+
+            for child in children:
+                results.append(child['slug'])
+
+                if 'children' in child:
+                    results.extend(fetch_children(child['children']))
+
+            return results
+
+        response = self._get(self.api_base_url + '/informationobjects/tree/' + resource_id)
+        tree = response.json()
+        return fetch_children(tree['children'])
```

Comment:
This is not going to work if the collection is empty, right? AtoM seems to omit the field instead of returning an empty list.